### DISCUSSION TITLE
Adding `%define parse.trace` in lrparser.y

### DISCUSSION THREAD

**Discussion by durwasa-chakraborty** [gajendra-iitm/starplat#7] [Ideas]
Hello

I am starting this thread regarding including the `%define parse.trace {}` construct in our `lrparser.y` file. @Barenya255  originally proposed this idea, and you can refer to the related PR [here](https://github.com/ashwinktpu/StarPlat/pull/80).

### Overview

The `%define parse.trace {}` is a Bison construct that aims to enhance the parser by providing detailed trace information during parsing. This can be highly beneficial for debugging and understanding the parser's behavior in more complex scenarios.

### Observations and Current Challenges

While attempting to integrate this feature, we encountered several challenges that caused regressions in our current codebase. We experienced issues when trying to run this on different operating system variants, including macOS and Pop!_OS. Despite multiple attempts, the feature did not work as expected across these systems. 

### Argument for Inclusion

Despite the regressions, I believe that the inclusion of `%define parse.trace {}` in our codebase would primarily benefit our project in the following ways:

1. **Enhanced Debugging:** The trace information will provide a clear insight into the parser's decisions, making identifying and fixing bugs easier.
2. **Improved Development Efficiency:** With better visibility into the parsing process, developers can quickly understand and resolve parsing issues, thus speeding up the development cycle.
3. **Educational Value:** Trace logs can serve as a valuable educational resource for new contributors or those who want to understand the internals of our parser.

### Proposed Changes

To ensure that this feature can be successfully integrated without causing disruptions, I propose the following steps:

1. **Cross-Platform Compatibility Testing:** Perform comprehensive testing on various operating systems to ensure compatibility. This includes setting up automated tests that can verify the functionality on macOS, Pop!_OS, and other major OS variants.
2. **Incremental Integration:** Introduce the `%define parse.trace {}` feature in stages, starting with a dedicated branch. This will allow us to monitor its impact and make necessary adjustments before merging it into the main codebase.
3. **Community Feedback:** Encourage community feedback during testing to identify and address any unforeseen issues.

cc: @Barenya255  , could you expand on how we should go about adding the changes? I look forward to hearing your thoughts and feedback on this proposal.

**Comment by Barenya255** [comment]
# YACC and POSIX portability.
Initially, I thought it could be the issue with POSIX / YACC portability, a specified requirement in the bison documentation. However, that does not seem to be the case. The generated codes, y.tab.h and y.tab.cc should compile and run.

# Possible issue with Makefile.
As the error occurs when the Makefile is linking, it could be an issue with either the Makefile or make itself (for macOS).

# Some checklists : 
- try to just generate the y.tab.cc and y.tab.h files using bison.
- if they generate compare these two generated files across an ubuntu machine and a mac machine. If they are different we could investigate further into bison's version and bison in macOS (arm).
- if the generated files are the same, the issue would be in the Makefile, we'll have to individually go through all the compilation commands to find what is causing the linkage to fail.

Additionally, snapshots of the error would be helpful.
